File: app/schemas/mcp.py
from datetime import timedelta
from pathlib import Path
from typing import Any, Literal, TypedDict

from pydantic import BaseModel


class StdioConnection(TypedDict):
    transport: Literal["stdio"]
    command: str
    """The executable to run to start the server."""
    args: list[str]
    """Command line arguments to pass to the executable."""
    env: dict[str, str] | None
    """The environment to use when spawning the process."""
    cwd: str | Path | None
    """The working directory to use when spawning the process."""
    encoding: str
    """The text encoding used when sending/receiving messages to the server."""


class SSEConnection(TypedDict):
    transport: Literal["sse"]
    url: str
    """The URL of the SSE endpoint to connect to."""
    headers: dict[str, Any] | None
    """HTTP headers to send to the SSE endpoint"""
    timeout: float
    """HTTP timeout"""
    sse_read_timeout: float
    """SSE read timeout"""


class StreamableHttpConnection(TypedDict):
    transport: Literal["streamable_http"]
    url: str
    """The URL of the endpoint to connect to."""
    headers: dict[str, Any] | None
    """HTTP headers to send to the endpoint."""
    timeout: timedelta
    """HTTP timeout."""
    sse_read_timeout: timedelta
    """How long (in seconds) the client will wait for a new event before disconnecting.
    All other HTTP operations are controlled by `timeout`."""
    terminate_on_close: bool
    """Whether to terminate the session on close"""


class WebsocketConnection(TypedDict):
    transport: Literal["websocket"]
    url: str
    """The URL of the Websocket endpoint to connect to."""


Connection = StdioConnection | SSEConnection | StreamableHttpConnection | WebsocketConnection


class MCPConnection(BaseModel):
    id: str
    connection: Connection
    name: str
    description: str | None = None

File: src/types/mcp.ts
// StdioConnection 接口
interface StdioConnection {
  transport: 'stdio';
  /** The executable to run to start the server. */
  command: string;
  /** Command line arguments to pass to the executable. */
  args: string[];
  /** The environment to use when spawning the process. */
  env?: Record<string, string>;
  /** The working directory to use when spawning the process. */
  cwd?: string;
  /** The text encoding used when sending/receiving messages to the server. */
  encoding: string;
}

// SSEConnection 接口
interface SSEConnection {
  transport: 'sse';
  /** The URL of the SSE endpoint to connect to. */
  url: string;
  /** HTTP headers to send to the SSE endpoint */
  headers?: Record<string, any>;
  /** HTTP timeout */
  timeout: number;
  /** SSE read timeout */
  sse_read_timeout: number;
}

// StreamableHttpConnection 接口
interface StreamableHttpConnection {
  transport: 'streamable_http';
  /** The URL of the endpoint to connect to. */
  url: string;
  /** HTTP headers to send to the endpoint. */
  headers?: Record<string, any>;
  /** HTTP timeout. */
  timeout: number;
  /**
   * How long (in seconds) the client will wait for a new event before disconnecting.
   * All other HTTP operations are controlled by `timeout`.
   */
  sse_read_timeout: number;
  /** Whether to terminate the session on close */
  terminate_on_close: boolean;
}

// WebsocketConnection 接口
interface WebsocketConnection {
  transport: 'websocket';
  /** The URL of the Websocket endpoint to connect to. */
  url: string;
}

export type AnyMCPConnection =
  | StdioConnection
  | SSEConnection
  | StreamableHttpConnection
  | WebsocketConnection;

export interface MCPConnection {
  id: string;
  name: string;
  description?: string;
  connection: AnyMCPConnection;
}
